core/endorser/endorser.go

```diff
@@ -472,86 +473,112 @@ func (e *Endorser) ProcessProposalSuccessfullyOrError(up *UnpackedProposal) (*pb
-			})
-		} else {
-			for collectionName, isRead := range collections {
+			}
+			if collection == "" { // the empty collection name here represents the public RWset
+				keyPolicies := policies.sbePolicies[chaincode]
+				if len(keyPolicies) > 0 {
+					// For simplicity, we'll always add the SBE policies to the public ChaincodeCall, and set the disregard flag if the chaincode policy is not required.
+					ccCall.KeyPolicies = keyPolicies
+					if !policies.requireChaincodePolicy(chaincode) {
+						ccCall.DisregardNamespacePolicy = true
+					}
+				} else if !policies.requireChaincodePolicy(chaincode) {
+					continue
+				}
+			} else {
 				// Since each collection in a chaincode could have different values of the NoPrivateReads flag, create a new Chaincode entry for each.
-				ccInterest.Chaincodes = append(ccInterest.Chaincodes, &pb.ChaincodeCall{
-					Name:            chaincode,
-					CollectionNames: []string{collectionName},
-					NoPrivateReads:  !isRead,
-				})
+				ccCall.CollectionNames = []string{collection}
+				ccCall.NoPrivateReads = !simResult.PrivateReads.Exists(chaincode, collection)
 			}
+			ccInterest.Chaincodes = append(ccInterest.Chaincodes, ccCall)
 		}
 	}
 
-	// add the key signature policies - at the moment they are grouped together in the first chaincode
-	// TODO implement DisregardNamespacePolicy hint in ChaincodeCall
-	if len(ccInterest.Chaincodes) > 0 {
-		for _, policyBytes := range simResult.KeySignaturePolicies {
-			policy, err := protoutil.UnmarshalSignaturePolicy(policyBytes)
-			if err != nil {
-				return nil, err
+	endorserLogger.Info("ccInterest", ccInterest)
+	return ccInterest, nil
+}
+
+type metadataPolicies struct {
+	// Map of SBE policies: namespace -> array of policies.
+	sbePolicies map[string][]*common.SignaturePolicyEnvelope
+	// Whether the chaincode/collection policy is required for endorsement: namespace -> collection -> isRequired
+	// Empty collection name represents the public rwset
+	// Each entry in this map represents a ChaincodeCall structure in the final ChaincodeInterest.  The boolean
+	// flag isRequired is used to control whether the DisregardNamespacePolicy flag should be set.
+	policyRequired map[string]map[string]bool
+}
+
+func parseWritesetMetadata(metadata ledger.WritesetMetadata) (*metadataPolicies, error) {
+	mp := &metadataPolicies{
+		sbePolicies:    map[string][]*common.SignaturePolicyEnvelope{},
+		policyRequired: map[string]map[string]bool{},
+	}
+	for ns, cmap := range metadata {
+		mp.policyRequired[ns] = map[string]bool{"": false}
+		for coll, kmap := range cmap {
+			// look through each of the states that were written to
+			for _, stateMetadata := range kmap {
+				if policyBytes, sbeExists := stateMetadata[pb.MetaDataKeys_VALIDATION_PARAMETER.String()]; sbeExists {
+					policy, err := protoutil.UnmarshalSignaturePolicy(policyBytes)
+					if err != nil {
+						return nil, err
+					}
+					mp.sbePolicies[ns] = append(mp.sbePolicies[ns], policy)
```

Comment:
Possibly.  I wanted to minimise the processing overhead of this function as much as possible since this gets invoked for all endorsers whereas it's only the first one that gets looked at by the gateway.  I intended the gateway to do the bulk of this type of processing.  If we find, in practice, that there are situations where large numbers of keys all have the same SBE policy, then could look to merge them in this function.